TMT enablement

jkonecny12 opened the pull request:
Enable TMT for Simpleline unit tests and smoke test.

This will be used by Packit and could be run locally.

TODO:
- [ ] Enable local execution with prepare step which has artifacts. (Right now it can't import simpleline).

jkonecny12 committed:
Add TMT smoke test

jkonecny12 committed:
Add unit tests to TMT

thrix commented on plans/smoke.fmf:
```diff
@@ -0,0 +1,9 @@
+summary: Run smoke test
+discover:
```
I believe `discover` steps does not need to be here. You are running a script in `execute` a script, thus discovered tests will be ignored I believe

thrix replied:
@psss right?

jkonecny12 replied:
Thanks for your review! Solved

thrix commented:
/packit test

jkonecny12 commented:
/packit test

poncovka reviewed (approved):
Looks good to me.

psss commented:
If you set the unit test `path` to the root of the git repo, `import simpleline` will work:

    summary: Run unit tests
    description: Run all unit tests for the Simpleline project.
    contact: Jiri Konecny <jkonecny@redhat.com>
    duration: 10m
    path: /
    test: ./tests/units/run_test.sh

However it still fails on other deps:

    tmt run -vvv plan --name basic
    ...
    ModuleNotFoundError: No module named 'gi'

Adding required python module into the plan will fix this:

    prepare:
        how: install
        package: python3-gobject-base

jkonecny12 commented:
> If you set the unit test `path` to the root of the git repo, `import simpleline` will work:
> 
> ```
> summary: Run unit tests
> description: Run all unit tests for the Simpleline project.
> contact: Jiri Konecny <jkonecny@redhat.com>
> duration: 10m
> path: /
> test: ./tests/units/run_test.sh
> ```
> 

Yes, I know but in that case the Packit will not test installed source but instead it will test always the source code. I want to create new plan which will be ignored by Packit and that will work for the local run. I just didn't had time to finish that yet.

Thanks a lot for the prepare info. You saved my time in debugging that ;).

jkonecny12 committed:
Move existing TMT plans under pr-testing directory

This configuration will work only for PR testing or gating.

jkonecny12 committed:
Add new TMT plan and split tests for local and pr-tests

PR-testing and Gating have to test unit-tests based on the installed
package, however, when we are running these tests in local then we want
to start the tests on source and not the installed package.

Create two plans to solve this issue.

jkonecny12 commented:
**UPDATED:**
* Split plans to be make a correct preparation and test execution for local and pr-tests.

jkonecny12 commented:
@psss What do you think, any space for improvement of my new configuration?

jkonecny12 commented:
@thrix any idea about when my TMT configuration may work on cruncher?

poncovka reviewed (commented):
I am sorry, but I don't understand it at all.

psss commented:
Looks good. Perhaps, if you would prefer to have both tests defined in a single file, you can do it like this:

```yaml
summary: Run unit tests
description: Run all unit tests for the Simpleline project.
contact: Jiri Konecny <jkonecny@redhat.com>
duration: 10m

/source:
    summary+: ' (source)'
    test: ./tests/units/run_test.sh
    path: /
    tag: local

/system:
    summary+: ' (system installed)'
    test: ./units/run_test.sh
    path: /tests
    tag: system
```

The extended summary is, of course, just an optional enhancement.

thrix commented:
> @thrix any idea about when my TMT configuration may work on cruncher?

Wanted to make last week, did not work out :/ .. so this week, I plan to work on it from tomorrow

jkonecny12 commented:
**UPDATED:**
* Updated based on @psss suggestion. Thanks a lot!

jkonecny12 commented:
jenkins, test this please

TomasTomecek commented:
/packit build

jkonecny12 commented:
/packit build

thrix commented:
Ah F33 has branched, I am fixing fedora-32 now.

thrix commented:
/packit test

jkonecny12 commented:
/packit test

thrix commented:
Sorry I messed up, deploying another fix

thrix commented:
/packit test

thrix commented:
fixed :)

jkonecny12 commented:
/packit test

jkonecny12 commented:
/packit build

jkonecny12 commented:
/packit build

poncovka reviewed (approved):
Looks good to me now. Thank you!

jkonecny12 commented:
Hmm, seems this is still blocked by Cruncher:
```
 No 'script' defined in the execute step. 
```

I hope it will be fixed soon :(.

thrix commented:
@jkonecny12 we are almost there, but we got blocked on getting [CentOS into AWS](https://url.corp.redhat.com/TFT-442) :/ Sorry for the delay

VladimirSlavik commented:
Do we really need to have all these files all over the repository? Isn't there some way to have them contained in .fmf, or tests/tmt, or such?

psss commented:
> Do we really need to have all these files all over the repository? Isn't there some way to have them contained in .fmf, or tests/tmt, or such?

The `.fmf` directory marks the root of the metadata tree (similar to what `.git` does) and it should be in the git repository root. As far are tests and plans are concerned, it is possible to choose arbitrary location for them. For example like this:

    tests/plans ... for L2 metadata
    tests/tests ... for L1 metadata

But storing test code & metadata under `tests` and plan configuration under `plans` seem to be a reasonable default.

> Also, it makes no sense to me that we're enabling TMT but all related filesystem bits are called FMF.

The `fmf` format is used to store metadata with support for inheritance and elasticity (you can look at it as an extended `yaml`) while `tmt` is the test metadata specification and tool for easily running them from your laptop:

 * https://fmf.readthedocs.io/
 * https://tmt.readthedocs.io/

Hope this helps.

jkonecny12 commented:
> @jkonecny12 we are almost there, but we got blocked on getting [CentOS into AWS](https://url.corp.redhat.com/TFT-442) :/ Sorry for the delay

Looking forward to that! Thanks for the update.

VladimirSlavik commented:
> As far are tests and plans are concerned, it is possible to choose arbitrary location for them.

And that would happen in some configuration file? (Just asking to understand.)

> But storing test code & metadata under `tests` and plan configuration under `plans` seem to be a reasonable default.

Well, both yes and no... I mean, from the testing point of view, certainly yes. But if I'm a wannabe contributor, open the repository and see the file listing, I see mostly stuff not really related to developing the thing. With some .hidden directories I can at least tell I can ignore them. Hope that explains why I'm asking.

> Hope this helps.

Thank you, it does!

jkonecny12 commented:
> > As far are tests and plans are concerned, it is possible to choose arbitrary location for them.
> 
> And that would happen in some configuration file? (Just asking to understand.)
> 
> > But storing test code & metadata under `tests` and plan configuration under `plans` seem to be a reasonable default.
> 
> Well, both yes and no... I mean, from the testing point of view, certainly yes. But if I'm a wannabe contributor, open the repository and see the file listing, I see mostly stuff not really related to developing the thing. With some .hidden directories I can at least tell I can ignore them. Hope that explains why I'm asking.
> 
> > Hope this helps.
> 
> Thank you, it does!

Hi @VladimirSlavik , please ask me directly. I think I'm able to answer most of your questions and it's much easier to discuss this on BJ/IRC than GH. ;)

martinpitt commented on plans/local/main.fmf:
```diff
@@ -0,0 +1,8 @@
+    discover+:
+        filter: tag:local
+    prepare:
+        - name: Install missing dependencies
```
This syntax is invalid. I copied that [in my own playground](https://github.com/martinpitt/python-simpleline/pull/1), and [filed a wrong bug](https://pagure.io/fedora-ci/general/issue/170) as I was misled by the failure output.

`prepare:` is a dict, not a list. See the [documentation](https://tmt.readthedocs.io/en/latest/spec/steps.html#prepare)

jkonecny12 replied:
Thanks a lot for your insight. Interesting, I'm pretty sure that worked locally :D

jkonecny12 replied:
At the end this syntax is supported, see:

https://github.com/psss/tmt/blob/master/plans/unit.fmf#L9

Seems that any stage could be specified like this to be able specify multiple different steps. However, it's not really required here.

Unfortunately, I can't find any sign about this in the documentation. Am I missing something @psss?

psss replied:
Yes, steps support multiple definitions. This is documented in the [multiple configs](https://tmt.readthedocs.io/en/latest/examples.html#multiple-configs) section and included in the [examples](https://github.com/psss/tmt/blob/master/examples/multiple/basic.fmf) as well. Were would you suggest to add an additional mention about this?

jkonecny12 replied:
@psss
Ohh, thanks a lot for the pointers. I totally missed the ``multiple configs`` section. I guess the problem is that it's just in the ``Examples`` section and not mentioned in the ``Specification part``. So maybe, it would be nice to add this to ``Metadata Specification`` section too with explanation that both variants are valid.

martinpitt commented:
FTR, I played around with that a bit, and https://github.com/martinpitt/python-simpleline/pull/1 now runs the tests correctly (except for a pylint version issue on older Fedoras, but that's not TMT's fault). Still requires some hacks, but these are said to go away with the TMT migration. So indeed let's wait with this for a bit until that happened. Until then, we have unit tests on github.

jkonecny12 commented:
**UPDATED:**
* Fixed bad syntax pointed by @martinpitt , thanks!

jkonecny12 commented:
/packit build

jkonecny12 commented:
/packit build
